### PR TITLE
[Component]: Input

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Fixes https://github.com/brave/leo/issues/54
Closes #79

### CODE REVIEW

**Comment by aguscruiz** [comment]
Couple comments:

- How about we use the text cursor on hover/active? 
- When you switch the input to have an icon, it gets a bit taller, it shouldn't change size. The right size is the one with no icon.

File: src/components/input/input.stories.svelte [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+<Story name="Default" />
+
+<Story name="Character Count" let:args>
+  <div style={getStyleFromArgs(args)}>
+    <Input {...getNonStyleArgs(args)} bind:value={characterCountValue}>
+      {args.label}
+      <div slot="extra">{characterCountValue.length}/100</div>
+      <div slot="right-icon">
+        <Icon name="send" />
+      </div>
```
Comment (AlanBreck) [inline src/components/input/input.stories.svelte:59]:
```suggestion
      <Icon name="send" slot="right-icon" />
```

File: src/components/formItem/formItem.svelte
Comment (AlanBreck) [inline src/components/formItem/formItem.svelte:1]: Possible improvement: can you remove the empty divs when no icons are provided?

Comment (fallaciousreasoning) [reply src/components/formItem/formItem.svelte:1]: Okay, so I had a quick play around with this - I think it actually makes things more complicated, as I need to:
1. Add another class for the central div, so we can apply the padding
2. Set display: flex; flex-direction: row; justify-content: stretch; to get the content to fill the div

So I've left this as is

**Comment by fallaciousreasoning** [comment]
@petemill mentioned we should look at why there are two variables for showing errors

**Comment by fallaciousreasoning** [comment]
Okay @petemill I've dug back into why we have two properties for determining whether we show errors

`hasErrors` indicates whether the component has any custom errors (i.e. `your password must contain the letter Z`)
`showErrors` determines whether the component should display the errors

We also have a required property, and we wanted to be able to manage that automatically. These are the states that we need to be able to handle:
1. A required input is empty, but the form has not been submitted (errors should not be shown)
2. A required input has text, has other errors, but has not been submitted (errors should not be shown)
3. A required input is empty and the form has been submitted (errors should be shown)
4. A required input has text, has other errors and has been submitted (errors should be shown)

This is how I was imagining it being used:

```tsx

const [errors, setErrors] = React.useState<string[]>([])
const [submitted, setSubmitted] = React.useState(false)

return <Input required showErrors={submitted} hasErrors={errors.length}>
    <ul slot="errors>{errors.map(e => <li>{e}</li>}</ul>
</Input>
```

**Comment by petemill** [comment]
@fallaciousreasoning it makes sense to me, that's how I've done it in the past too. I just thought there were 3 properties for some reason!

**Review by petemill — APPROVED** [review]
Looks great, I think it was the errors slot that confused me as a third "prop" but it makes sense to have it separated and styleable.

**Comment by petemill** [comment]
@mattwilliams85 noticed lack of `placeholder` prop. Seems neccessary. We can do in this PR before merging, or a follow-up

**Comment by petemill** [comment]
Should other input props be passed on to the `<input`, especially aria? What about number related properties (min, max, regex, etc)?

**Pull request merged**